### PR TITLE
Back 749 - Register from a cursor state

### PR DESCRIPTION
**Pull request by nukky94**
## What is this about?
Register from a block height cursor state in order to support: 
- reseed (sync from beginning)
- reorg (sync from a valid block height)

## Cute picture of lama
B9723377337Z 1_20200504173403_000+GBAFVB9EP 1-0

### CODE REVIEW

File: account-manager/src/main/scala/co/ledger/lama/manager/Queries.scala [hunk 1]
```diff
@@ -17,7 +17,10 @@ import scala.concurrent.duration.FiniteDuration
 
 object Queries {
 
-  def fetchCandidateEvents(coinFamily: CoinFamily, coin: Coin): Stream[ConnectionIO, SyncPayload] =
+  def fetchCandidateEvents(
+      coinFamily: CoinFamily,
+      coin: Coin
+  ): Stream[ConnectionIO, CandidateSyncEvent] =
```
Comment (vbergeron-ledger) [inline account-manager/src/main/scala/co/ledger/lama/manager/Queries.scala:23]: format

Comment (nukky94) [reply account-manager/src/main/scala/co/ledger/lama/manager/Queries.scala:23]: actually, it was auto formatted by scalafmt :D

File: account-manager/src/main/protobuf/service.proto [hunk 2]
```diff
@@ -6,21 +6,39 @@ option java_package = "co.ledger.lama.manager.protobuf";
...
 
-message AccountInfoRequest {
+message RegisterAccountRequest {
     string extendedKey = 1;
     CoinFamily coinFamily = 2;
     Coin coin = 3;
     uint64 syncFrequency = 4;
+    oneof cursor {
+        BlockHeightState blockHeight = 5;
+    }
```
Comment (RomanWlm) [inline account-manager/src/main/protobuf/service.proto:20]: Uniontype with one value is a bit weird. Is this for future enhancement like cursor can be a hash or something else ?

File: account-manager/src/main/protobuf/service.proto [hunk 3]
```diff
@@ -6,21 +6,39 @@ option java_package = "co.ledger.lama.manager.protobuf";
...
 }
 
-message AccountInfoResult {
+message BlockHeightState {
+    uint64 state = 1;
```
Comment (RomanWlm) [inline account-manager/src/main/protobuf/service.proto:24]:
```suggestion
    uint64 height = 1;
```
Extravagant naming proposal :)

File: account-manager/src/main/scala/co/ledger/lama/manager/Service.scala [hunk 4]
```diff
@@ -101,28 +107,33 @@ class Service(val db: Transactor[IO], val coinConfigs: List[CoinConfig])
...
+                UuidUtils.uuidToBytes(event.syncId)
               )
             )
       }
     } yield result
   }
 
+  def cursorToJson(cursor: protobuf.RegisterAccountRequest.Cursor): Json = {
+    if (cursor.isBlockHeight)
+      Json.obj("blockHeight" -> Json.fromLong(cursor.blockHeight.map(_.state).getOrElse(0L)))
```
Comment (RomanWlm) [inline account-manager/src/main/scala/co/ledger/lama/manager/Service.scala:134]:
```suggestion
      Json.obj("blockHeight" -> Json.fromLong(cursor.blockHeight.map(_.height).getOrElse(0L)))
```
If previous has been taken

Comment (RomanWlm) [inline account-manager/src/main/scala/co/ledger/lama/manager/Service.scala:134]: height is an option ? getOrElse is a bit confusing here

File: account-manager/src/main/protobuf/service.proto (hunk 2 above)
Comment (nukky94) [reply account-manager/src/main/protobuf/service.proto:20]: Exactly, it's more resilient for future enhancement (in other blockchain, we could have a different kind of cursor for example)

File: account-manager/src/main/scala/co/ledger/lama/manager/Service.scala (hunk 4 above)
Comment (nukky94) [reply account-manager/src/main/scala/co/ledger/lama/manager/Service.scala:134]: Actually, this is due to the generated scala code from scalapb:
```
  sealed trait Cursor extends _root_.scalapb.GeneratedOneof {
    def isEmpty: _root_.scala.Boolean = false
    def isDefined: _root_.scala.Boolean = true
    def isBlockHeight: _root_.scala.Boolean = false
    def blockHeight: _root_.scala.Option[co.ledger.lama.manager.protobuf.BlockHeightState] = _root_.scala.None
  }
```

`oneOf` means union type, and it is not supported by scala, so the behaviour is to have multiple optional field and check the corresponding boolean.

File: account-manager/src/main/protobuf/service.proto (hunk 2 above)
Comment (RomanWlm) [reply account-manager/src/main/protobuf/service.proto:20]: Perfect !

File: account-manager/src/main/scala/co/ledger/lama/manager/Service.scala (hunk 4 above)
Comment (RomanWlm) [reply account-manager/src/main/scala/co/ledger/lama/manager/Service.scala:134]: Clear, make sense. 
Then if we have a None here this is an error case ? 
Asking myself if we should fail frankly or not..

Comment (nukky94) [reply account-manager/src/main/scala/co/ledger/lama/manager/Service.scala:134]: Yes, None is an error case but it couldn't happen if you checked correctly flags:
```
if (cursor.isBlockHeight) cursor.blockHeight.get // should never fail
else if (cursor.isBlockHash) cursor.blockHash.get // should never fail
else if .... // and so on
```

**Pull request merged**